zkevm-circuits/src/mpt_circuit/storage_leaf/leaf_value.rs

```diff
@@ -100,890 +101,159 @@ Note that there are no checks for example for the root as the constraints to ens
+            } elsex {
+                let value_rlc = s_main.bytes(meta, 0).rlc(&r);
+                let leaf_rlc = (0.expr(), mult_prev.expr()).rlc_chain([a!(s_main.rlp1), a!(s_main.rlp2), value_rlc.expr()].rlc(&r));
+                (value_rlc, leaf_rlc)
+            }};
+            require!(value_rlc => new_value_rlc);
+            require!(leaf_rlc => leaf_rlc.prev() + leaf_rlc_part);
+
+            // To enable external lookups we need to have the key and the previous/current
+            // value on the same row.
             if !is_s {
-                let is_wrong_leaf = meta.query_advice(
-                    s_main.rlp1,
-                    Rotation(LEAF_NON_EXISTING_IND - LEAF_VALUE_C_IND),
-                );
-                constraints.push((
-                    "is_wrong_leaf is bool",
-                    get_bool_constraint(q_enable.clone(), is_wrong_leaf.clone()),
-                ));
-
-                let is_non_existing_storage_proof =
-                    meta.query_advice(proof_type.is_non_existing_storage_proof, Rotation::cur());
-
-                // Note: some is_wrong_leaf constraints are in this config because
-                // leaf_non_existing config only triggers constraints for
-                // non_existing_storage proof (see q_enable).
-                constraints.push((
-                    "is_wrong_leaf needs to be 0 when not in non_existing_account proof",
-                    q_enable * (one.clone() - is_non_existing_storage_proof) * is_wrong_leaf,
-                ));
+                require!(a!(accs.key.mult) => a!(accs.key.rlc, rot_key));
+                require!(a!(value_prev) => value_rlc.prev());
+                require!(a!(value) => value_rlc);
             }
 
-            constraints
+            // Get the number of bytes used by the value
+            let value_num_bytes = matchx! {
+                is_short => 1.expr(),
+                is_long => 1.expr() + get_num_bytes_short(a!(s_main.rlp2)),
+            };
+
+            // If `is_modified_node_empty = 1`, which means an empty child, we need to
+            // ensure that the value is set to 0 in the placeholder leaf. For
+            // example when adding a new storage leaf to the trie, we have an empty child in
+            // `S` proof and non-empty in `C` proof.
+            ifx! {branch.contains_placeholder_leaf(meta, is_s) => {
+                require!(a!(s_main.rlp1) => 0);
+            }}
+
+            // Make sure the RLP encoding is correct.
+            // storage = [key, value]
+            let num_bytes = storage.num_bytes(meta);
+            // TODO(Brecht): modify the witness for empty placeholder leafs to have valid
+            // RLP encoding
+            ifx! {not!(branch.contains_placeholder_leaf(meta, is_s)) => {
+                let key_num_bytes = storage.num_bytes_on_key_row(meta);
+                require!(num_bytes => key_num_bytes.expr() + value_num_bytes.expr());
+            }};
```

Comment:
As above for the `leaf_key` - fixed in witness generator and the circuit code updated in https://github.com/Brechtpd/zkevm-circuits/pull/6/commits/0e23861ebd479cb5b62f29d01566c4c79846e203.